### ISSUE TITLE
DB index for deleted

### ISSUE THREAD

**Issue by tuky** [makinacorpus/django-safedelete#153]
Would you welcome a PR to change `SafeDeleteModel.deleted` into something like this?

```python
deleted = models.DateTimeField(editable=False, null=True, db_index=SAFE_DELETED_DB_INDEX)
```

where `SAFE_DELETED_DB_INDEX` will be optionally configurable via django settings (default to `False` for backwards compatibility).

**Comment by Gagaro** [comment]
It might be interesting. Did you have some performance issue without the index?

You can already add this index in the model meta:

```python
class Meta:
    indexes = [Index(fields=('deleted',))]
```

Do you need to have the date indexed or only the null value?

It might be more interesting to only index null deleted, something like:

```python
class Meta:
    indexes = [Index(fields=('deleted',), name='%(app_label)s_%(class)s_not_deleted', condition=Q(deleted__isnull=True))]
```

Then again, this index might not be used if you have other condition in your queries.

**Comment by tuky** [comment]
Thanks for response, great input!

> Did you have some performance issue without the index?

Exactly, e.g. in the django admin listing only not deleted objects is taking a while for big tables.

> Do you need to have the date indexed or only the null value?

Actually only the null value, however we're on MySQL and cannot make use of conditional indexes :cry:

Overall I'm thinking if this was such a great addition to the code. Right now we get away by always using our own custom sub class:

```python
class SafeDeleteDBIndexModel(SafeDeleteModel):
    deleted = models.DateTimeField(editable=False, null=True, db_index=True)

    class Meta:
        abstract = True
```

**Comment by Gagaro** [comment]
> e.g. in the django admin listing only not deleted objects is taking a while for big tables.

That's actually a use case pretty much everyone using this will have. I'm unsure about if creating an index by default would be warranted. It could be pretty useful.

I don't know about the difference in size between the index with or without the condition, but we could create the index with the condition even for MySQL, it would just be ignored.

**Comment by panhaoyu** [comment]
Nice work, and I think the default index is useful.

For a table with about 20K lines, just a simple query will cause about 5s, which is not acceptable, and 20K is not a large dataset.

As a user, I want to install the lib and do no harm to our previous work, however, the performance does harm the work, and I have to dive into the source code to find the index. I think making it default is more intuitional.

Anyway, thanks for your work, nice lib!

**Comment by Gagaro** [comment]
I ended up adding a normal index because there was an issue when trying to use the dynamic field name (defined by the settings) in the Meta class (maybe because settings are not set yet at this time of Django setup?).

**Comment by tuky** [comment]
great, thanks for the change! i would not expect the settings to fail. they're already used in https://github.com/makinacorpus/django-safedelete/blob/master/safedelete/config.py aren't they?

**Comment by Gagaro** [comment]
Yes, but it was still 'deleted' for the index and the one defined in the setting for the field name in the tests. I'm not sure why.

**Comment by ganiserb** [comment]
Hello!

@Gagaro 

Our team just stumbled upon an issue with this feature. We tested this in a PostgreSQL DB v12.14 with several models, some of which have a couple million records.

Our DB server reached max load and died under normal usage. Once we deleted the auto-created `btree` indexes for the `deleted` fields we could get the website to work again.

99% of all the rows in the DB have a `NULL` in the `deleted` column. Any clues as to why the performance would be impacted like this? We suspect it is because the index lookup added extra overhead even though it wasn't useful, but I don't buy it.

I am thinking of submitting a PR to make the index parameter configurable. I'd welcome some feedback.

Thanks!

**Comment by ganiserb** [comment]
I'm going to try to get as much info as I can for some time this week.

Unfortunately, we had to manually remove the indexes from the production DB where the issue became super evident. So I can't do `EXPLAIN ANALYZE` in production with the indexes at the moment, but I can generate a huge test dataset in a staging environment and we can compare:

- Queries in the current production DB without indexes.
- Queries in the large test dataset **with** the indexes.

**Comment by Gagaro** [comment]
Don't forget to `VACUUM ANALYZE` after generating the data to be sure that the planner is going to do its job properly.

**Comment by ganiserb** [comment]
@Gagaro Ok, here is the rundown: We have a complex SQL view that we access through an unmanaged django model. We use this to present all the relevant information to the user for a certain date range. This queryset can be accessed (and filtered) through an API. After we released the changes that included django-safedelete in many of our models, this endpoint started making our DB CPU usage to skyrocket until it just became unresponsive.

Visually:
<img width="886" alt="image" src="https://github.com/makinacorpus/django-safedelete/assets/3132593/c187a7af-7d97-40d6-b338-b30cbaef2b42">

So, here are the plans used by the DB for executing the query that makes up that SQL View. I ran the `EXPLAIN ANALYZE` of the query in production as it stands today (**without** the `deleted` field indices) and [this is what I got](https://pastebin.com/mzMvJ9fG).

Then I cloned the whole production DB in a new server (using `pg_dump` and `pg_restore`). I created new migrations to recreate the indices. I ran `VACUUM ANALYZE` in this copy, and then I ran `EXPLAIN ANALYZE` for the same SQL view query **with the indices**. [Here is the plan for that](https://pastebin.com/9FKrGZ1H).

The execution of the query WITH the indices took ~40 seconds longer.
